### PR TITLE
add MoE codes on a single card (experimental version for dy2sta)

### PR DESCRIPTION
**Pull request by SylarTiaNII**
HOW TO USE:
  1. cd examples/transformer/models/GPT/
  2. mkdir data & cd data & wget https://paddlenlp.bj.bcebos.com/models/transformers/gpt2/dataset/my-gpt2_text_document_ids.npz & cd ..
  3. python pretrain_moe/run.py -c 'pretrain_moe/configs/pretrain_moe_gpt_345M_single_card.yaml'

NOTE: This version leaves several HACK because of lacking OP,  including but not limited to:
  1. Currently only supports topk=1. (need Rsample OP)
  2. Currently moe_train_capacity_factor&moe_eval_capacity_factor have to equal to num_experts. (blame to paddle.put_along_axis_)

### CODE REVIEW

File: examples/transformer/models/GPT/pretrain_moe/configs/pretrain_gpt_base.yaml [hunk 1]
```diff
@@ -0,0 +1,93 @@
+Global:
```
Comment (haohongxiang) [inline examples/transformer/models/GPT/pretrain_moe/configs/pretrain_gpt_base.yaml]: 建议更名为`pretrain_moe_base.yaml`

File: examples/transformer/models/GPT/pretrain_moe/configs/pretrain_moe_gpt_345M_single_card.yaml [hunk 2]
```diff
@@ -0,0 +1,59 @@
+_base_: ./pretrain_gpt_base.yaml
```
Comment (haohongxiang) [inline examples/transformer/models/GPT/pretrain_moe/configs/pretrain_moe_gpt_345M_single_card.yaml]: 同上，建议更名为`pretrain_moe_base.yaml`

File: examples/transformer/models/GPT/pretrain_moe/export.py [hunk 3]
```diff
@@ -0,0 +1,103 @@
+# Copyright (c) 2022 PaddlePaddle Authors. All Rights Reserved.
```
Comment (haohongxiang) [inline examples/transformer/models/GPT/pretrain_moe/export.py]: `export.py`有验证过吗？没有的话先删掉

File: examples/transformer/models/GPT/pretrain_moe/impls.py [hunk 4]
```diff
@@ -0,0 +1,247 @@
+# Copyright (c) 2022 PaddlePaddle Authors. All Rights Reserved.
```
Comment (haohongxiang) [inline examples/transformer/models/GPT/pretrain_moe/impls.py:1]: `impls.py`中有很多代码是可以复用`GPT/pretrain/impls.py`的，建议`import`进来，不需要再写一份

File: examples/transformer/models/GPT/pretrain_moe/impls.py [hunk 5]
```diff
@@ -0,0 +1,247 @@
...
+from ppfleetx.data.tokenizers import GPTTokenizer, GPTChineseTokenizer
+from ppfleetx.models.language_model.gpt.dygraph.sequence_parallel_utils import register_sequence_parallel_allreduce_hooks
+
+MODEL_CLASSES = {
+    "GPT": (GPTTokenizer, "gpt2"),
+    "GPT-cn": (GPTChineseTokenizer, "gpt-cpm-large-cn"),
+}
+
+
+def _get_model_size(l, h, v, s):
```
Comment (haohongxiang) [inline examples/transformer/models/GPT/pretrain_moe/impls.py]: 另写一个针对MoE模型参数量计算的function，不能复用GPT模型

File: ppfleetx/models/language_model/gpt/dygraph/single_model.py [hunk 6]
```diff
@@ -431,18 +438,34 @@ def __init__(self,
...
+        #         ExpertLayer(d_model, dim_feedforward)
+        #         for e in range(self.num_experts)
+        #     ])
+
+        #     self.moe_mlp = MoELayer(
+        #         d_model=d_model,
+        #         experts=experts_list,
+        #         gate=self.gate,
+        #         top_k=self.top_k,
+        # recompute_interval=int(self.use_recompute))
```
Comment (haohongxiang) [inline ppfleetx/models/language_model/gpt/dygraph/single_model.py]: 没用的注释请删掉

File: examples/transformer/models/GPT/pretrain_moe/configs/pretrain_gpt_base.yaml (hunk 1 above)
Comment (SylarTiaNII) [reply examples/transformer/models/GPT/pretrain_moe/configs/pretrain_gpt_base.yaml]: done

File: examples/transformer/models/GPT/pretrain_moe/configs/pretrain_moe_gpt_345M_single_card.yaml (hunk 2 above)
Comment (SylarTiaNII) [reply examples/transformer/models/GPT/pretrain_moe/configs/pretrain_moe_gpt_345M_single_card.yaml]: done

File: examples/transformer/models/GPT/pretrain_moe/impls.py (hunk 4 above)
Comment (SylarTiaNII) [reply examples/transformer/models/GPT/pretrain_moe/impls.py:1]: 后面完成了moe动手&静半功能之后，现有的GPT/pretrain文件夹需要被现在的GPT/pretrain_moe文件夹替换掉，现在只能临时出两套代码了。

File: examples/transformer/models/GPT/pretrain_moe/export.py (hunk 3 above)
Comment (SylarTiaNII) [reply examples/transformer/models/GPT/pretrain_moe/export.py]: done

File: ppfleetx/models/language_model/gpt/dygraph/single_model.py (hunk 6 above)
Comment (SylarTiaNII) [reply ppfleetx/models/language_model/gpt/dygraph/single_model.py]: done

File: examples/transformer/models/GPT/pretrain_moe/impls.py (hunk 5 above)
Comment (SylarTiaNII) [reply examples/transformer/models/GPT/pretrain_moe/impls.py]: done

File: examples/transformer/models/GPT/pretrain_moe/impls.py (hunk 4 above)
Comment (haohongxiang) [reply examples/transformer/models/GPT/pretrain_moe/impls.py:1]: 好的

**Review by haohongxiang — APPROVED** [review]
LGTM

**Pull request merged**